### DISCUSSION TITLE
I have noticed an issue with handle() that is actually in cfltk

### DISCUSSION THREAD

**Discussion by jgriffitts** [fltk-rs/fltk-rs#1501] [General]
I have noticed and fixed an issue that manifests in fltk-rs but is actually in [cfltk](https://github.com/MoAlyousef/cfltk).

Specifically, when the handle() method is used to set up an event handler for a widget, the resulting algorithm incorrect.

What should happen is that the local handler (as defined by the handle() method) should be called to process each event. If it chooses not to consume the event, it will return false, which causes the event to then be passed to the default handler defined for the corresponding fltk widget.

The bug is that _both_ handlers were always being called and in the wrong order: the default handler first, and then the local handler was called regardless of whether the default handler consumed the event.. This meant that the local handler couldn't override any of the default actions.

I have submitted a [PR](https://github.com/MoAlyousef/cfltk/pull/231) that fixes this in cfltk.

A side effect of this fix showed up in my own work: the default handler calls `redraw()` for FOCUS/UNFOCUS events and I had failed to do so in my local handler. With the change, `redraw()` was not called when it needed to be, the window updated in a confusingly erratic way. This seemed like a bug in the change, but only represented a top-level bug _revealed by_ the change.

For reference, here's the existing incorrect implementation:
```c++
    int handle(int event) override {
        int ret = T::handle(event);
        int local = 0;
        if (inner_handler) {
            local = inner_handler(this, event, ev_data_);
            return ret | local;
        } else {
            return ret;
        }
    }
```
Here is the version that I have submitted:
```c++
    int handle(int event) override {
        if (inner_handler) {
            if (inner_handler(this, event, ev_data_))
                return 1;       // Local handler consumed the event
        }
        return T::handle(event);    // Try the default handler
    }
```

**Comment by MoAlyousef** [comment]
Thank you for the PR.
I would say that almost always you would need the base handler to be invoked.
For example, if you try the following program with the proposed change:
```rust
use fltk::{app, button::Button, frame::Frame, prelude::*, window::Window, enums::*};

fn main() {
    let app = app::App::default();
    let mut wind = Window::default().with_size(400, 300);
    let mut frame = Frame::default().with_size(200, 100).center_of(&wind);
    let mut but = Button::new(160, 210, 80, 40, "Click me!");
    wind.end();
    wind.show();

    but.set_callback(move |_| frame.set_label("Hello world"));
    but.handle(|b, ev| match ev {
        Event::Push => {
            println!("pushed");
            true
        }
        _ => false,
    });
    app.run().unwrap();
}
```
The Push event will not trigger the callback, also the button animations will be gone since the Fl_Button::handle(FL_PUSH) is no longer called.

A solution to the order and to whether the base::handle() is called or not can be done like the overriden draw method, which has a conditional:
```cpp
     if (super_draw && super_draw_first)
            T::draw();
```
Then another in the end of the method.

And these are modified on the Rust side using:
```rust
    fn super_draw(&mut self, flag: bool);
    fn super_draw_first(&mut self, flag: bool);
```

As such, super_handle() determines whether the base's handler gets called or not, and the super_handle_first() determines whether it's done before or after the local handler.

This wasn't added since for most cases I think the super handler should be invoked first rather than after. But I'm ok with adding such modification.

**Reply by MoAlyousef** [reply]
I've added super_handle and super_handle_first. Changing the order of the base handler can be done using: `some_widget.super_handle_first(false);`.
To override the handler completely: super_handle(false) can be used.

**Comment by jgriffitts** [comment]
Perhaps I'm confused, but I don't think this super_handle_first change works. For reference, here's the new code:
```
    int handle(int event) override {
        int ret = 0;
        int local = 0;
        if (super_handle && super_handle_first)
            ret = T::handle(event);
        if (inner_handler) {
            local = inner_handler(this, event, ev_data_);
        }
        if (super_handle && !super_handle_first)
            ret = T::handle(event);
        return ret | local;
    }
```
With this structure, either `T::handle` is _never_ called for _any_ event, or it is _always_ called, either before or after the `inner_handler`.
The behavior I'm needing is to allow `inner_handler` to consume _some_ events, by returning TRUE which ends the event-handling process, and `T::handle` is not executed. Other events pass through to `T::handle` and if it returns FALSE in its turn, the events are passed on up the handler chain per fltk event propagation rules.

This could be patched by qualifying the last `if` statement with `local` but that is getting really cumbersome. I'd like to suggest getting rid of the `super_handle` flag and just using `super_handle_first` to switch between your preferred behavior and the standard fltk event-handling strategy, as follows:
```
    // Widget event handler. Manages execution of both event handlers:
    //   inner_handler is the handler defined by end-user program, and may not exist
    //   T::handle is the built-in handler in the fltk widget
    int handle(int event) override {
        if (super_handle_first) {   
            // Both handlers always executed, T::handle executes first
            int ret = T::handle(event);
            if (inner_handler) {
                int local = inner_handler(this, event, ev_data_);
                return ret | local;
            } else {
                return ret;
            }
        } else { 
            // inner_handler executes first. T::handle only executes if inner_handler returns false
            if (inner_handler) {
                if (inner_handler(this, event, ev_data_))
                    return 1; // Local handler consumed the event
            }
            return T::handle(event); // Try the default handler
        }
    }
```

**Reply by MoAlyousef** [reply]
Ok I see. That seems reasonable, I mean a flag which stops propagation to base once the specified event is handled. Would you like to open a PR?
